### PR TITLE
[feat][misc] PIP-264: Implement topic lookup metrics using OpenTelemetry

### PR DESCRIPTION
**Pull request by dragosvictor**
[PIP-264](https://github.com/apache/pulsar/blob/master/pip/pip-264.md)

### Motivation

Part of PIP-264 related work to convert metrics to OpenTelemetry format. This PR handles topic lookup metrics, as currently defined in the [documentation](https://pulsar.apache.org/docs/next/reference-metrics/#topic-lookup-metrics). 

Note that metric `pulsar_broker_load_manager_bundle_assignment` fits better into the Load Balancing metrics section, so is not included in this work.

### Modifications

Added metrics as described below:

| Old metric name | New metric details | |
|---|---|---|
| `pulsar_broker_lookup` | Name | `pulsar.broker.lookup.request.duration` |
| | Type | Histogram |
| | Description | The duration of topic lookup requests (either binary or HTTP)
| | Source | Internal OpenTelemetry implementation
| | Attribute Set | `pulsar.lookup.response=failure` replaces `pulsar_broker_lookup_failures`
| | Attribute Set | `pulsar.lookup.response=broker` replaces `pulsar_broker_lookup_answers`
| | Attribute Set | `pulsar.lookup.response=redirect` replaces `pulsar_broker_lookup_redirects`
| | Cardinality | 3 per broker (one per attribute set)
| `pulsar_broker_lookup_pending_requests` | Name | `pulsar.broker.topic.lookup.operation.pending.usage` |
| | Type | ObservableLongUpDownCounter |
| | Description | The number of pending lookup operations in the broker. When it reaches threshold \"maxConcurrentLookupRequest\" defined in broker.conf, new requests are rejected.
| | Source | Call to `BrokerService#getPendingLookupRequest`
| | Attributes | (none)
| | Cardinality | 1 per broker
| (new metric) | Name | `pulsar.broker.topic.lookup.operation.pending.limit` |
| | Type | ObservableLongUpDownCounter |
| | Description | The maximum number of pending lookup operations in the broker. Equal to \"maxConcurrentLookupRequest\" defined in broker.conf.
| | Source | Call to `ServiceConfiguration#getMaxConcurrentLookupRequest`
| | Attributes | (none)
| | Cardinality | 1 per broker
| `pulsar_broker_topic_load_pending_requests` | Name | `pulsar.broker.topic.load.operation.pending.usage` |
| | Type | ObservableLongUpDownCounter |
| | Description | The number of pending topic load operations in the broker. When it reaches threshold \"maxConcurrentTopicLoadRequest\" defined in broker.conf, new requests are rejected.
| | Source | Call to `BrokerService#getPendingTopicLoadRequests`
| | Attributes | (none)
| | Cardinality | 1 per broker
| (new metric) | Name | `pulsar.broker.topic.load.operation.pending.limit` |
| | Type | ObservableLongUpDownCounter |
| | Description | The maximum number of pending topic load operations in the broker. Equal to \"maxConcurrentTopicLoadRequest\" defined in broker.conf.
| | Source | Call to `ServiceConfiguration#getMaxConcurrentTopicLoadRequest`
| | Attributes | (none)
| | Cardinality | 1 per broker

Moved instantiation of `PulsarBrokerOpenTelemetry` object to `PulsarService` constructor, as it is needed by the `NamespaceService` and `BrokerService`. The OpenTelemetry standard recommends setting up the SDK as early as possible, so this suits us well.

Changed `PulsarTestContext` to allow OpenTelemetry metrics validation as part of the tests. 

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

*(example:)*
  - Extended existing lookup tests `BrokerServiceLookupTest#testMultipleBrokerLookup` and `BrokerServiceLookupTest#testLookupConnectionNotCloseIfGetUnloadingExOrMetadataEx` to cover the metric validation

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [x] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/dragosvictor/pulsar/pull/11

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 1]
```diff
@@ -146,18 +148,24 @@ public class NamespaceService implements AutoCloseable {
 
     private final RedirectManager redirectManager;
 
-
+    /** @deprecated by {@link #lookupRedirectsCounter} */
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:151]: Should we consider adding an ad-hoc Java annotation here. It would make our life easier when we'll remove these metrics.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 2]
```diff
@@ -205,16 +232,21 @@ public CompletableFuture<Optional<LookupResult>> getBrokerServiceUrlAsync(TopicN
                 });
 
         future.thenAccept(optResult -> {
-            lookupLatency.observe(System.nanoTime() - startTime, TimeUnit.NANOSECONDS);
+            var latencyNs = System.nanoTime() - startTime;
+            lookupLatency.observe(latencyNs, TimeUnit.NANOSECONDS);
+            lookupLatencyHistogram.record(latencyNs / 1_000_000_000.0);
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:237]: Should we add a utility method to convert nanos to seconds (with double)? It might come up multiple ways

**Review by merlimat — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 1 above)
Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:151]: Yes, something like `@PulsarDeprecatedMetric`. It would easier later on to go and find all of them

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 3]
```diff
@@ -146,18 +150,26 @@ public class NamespaceService implements AutoCloseable {
 
     private final RedirectManager redirectManager;
 
-
+    @PulsarDeprecatedMetric(newMetricName = "pulsar.broker.lookup.redirect")
```
Comment (merlimat) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:153]: Nice idea to put new metric here! We could use this to automatically create a conversion table

**Review by asafm — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java [hunk 4]
```diff
@@ -312,6 +313,8 @@ public PulsarService(ServiceConfiguration config,
         TransactionBatchedWriteValidator.validate(config);
         this.config = config;
 
+        this.openTelemetry = new PulsarBrokerOpenTelemetry(config);
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/PulsarService.java:316]: This brings me to the comment I made in that section. We shouldn't replace variables set at the constructor. We should provide them to the constructor.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 5]
```diff
@@ -205,16 +236,21 @@ public CompletableFuture<Optional<LookupResult>> getBrokerServiceUrlAsync(TopicN
                 });
 
         future.thenAccept(optResult -> {
-            lookupLatency.observe(System.nanoTime() - startTime, TimeUnit.NANOSECONDS);
+            var latencyNs = System.nanoTime() - startTime;
+            lookupLatency.observe(latencyNs, TimeUnit.NANOSECONDS);
+            lookupLatencyHistogram.record(MetricsUtil.convertToSeconds(latencyNs, TimeUnit.NANOSECONDS));
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: Ohhhhhhhh. Nice.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 6]
```diff
@@ -403,15 +411,41 @@ public BrokerService(PulsarService pulsar, EventLoopGroup eventLoopGroup) throws
...
 
         this.pendingTopicLoadRequests = ObserverGauge.build(
-                "pulsar_broker_topic_load_pending_requests", "-")
-                .supplier(() -> pulsar.getConfig().getMaxConcurrentTopicLoadRequest()
-                        - topicLoadRequestSemaphore.get().availablePermits())
+                        "pulsar_broker_topic_load_pending_requests", "-")
+                .supplier(this::getPendingTopicLoadRequests)
                 .register();
+        this.pendingTopicLoadRequestsCounter = pulsar.getOpenTelemetry().getMeter()
+                .gaugeBuilder("pulsar.broker.topic.load.pending.request.usage")
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: Good point. Yes.
@codelipenghui WDYT?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 7]
```diff
@@ -241,8 +243,15 @@ public class BrokerService implements Closeable {
...
     protected final AtomicReference<Semaphore> topicLoadRequestSemaphore;
 
+    @PulsarDeprecatedMetric(newMetricName = "pulsar.broker.lookup.pending.request.usage")
     private final ObserverGauge pendingLookupRequests;
+    private final ObservableLongGauge pendingLookupRequestsCounter;
+    private final ObservableLongGauge pendingLookupRequestsLimit;
+
+    @PulsarDeprecatedMetric(newMetricName = "pulsar.broker.topic.load.pending.request.usage")
     private final ObserverGauge pendingTopicLoadRequests;
+    private final ObservableLongGauge pendingTopicLoadRequestsCounter;
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: Not sure I understand.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/PulsarBrokerOpenTelemetry.java [hunk 8]
```diff
@@ -34,10 +37,17 @@ public class PulsarBrokerOpenTelemetry implements Closeable {
...
+    }
+
+    @VisibleForTesting
+    public PulsarBrokerOpenTelemetry(ServiceConfiguration config,
+                                     Consumer<AutoConfiguredOpenTelemetrySdkBuilder> builderCustomizer) {
         openTelemetryService = OpenTelemetryService.builder()
                 .clusterName(config.getClusterName())
                 .serviceName(SERVICE_NAME)
                 .serviceVersion(PulsarVersion.getVersion())
+                .sdkBuilderConsumer(builderCustomizer)
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/PulsarBrokerOpenTelemetry.java:50]: Sorry, I misspelled.
I meant: `sdkBuilderConsumer` --> `sdkBuilderCustomizer`

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/PulsarTestContext.java [hunk 9]
```diff
@@ -727,11 +733,24 @@ protected void initializePulsarServices(SpyConfig spyConfig, Builder builder) {
...
+            PulsarBrokerOpenTelemetry pulsarBrokerOpenTelemetry;
+            if (builder.enableOpenTelemetry) {
+                var reader = InMemoryMetricReader.create();
+                pulsarBrokerOpenTelemetry = new PulsarBrokerOpenTelemetry(builder.config, builderCustomizer -> {
+                    builderCustomizer.addMeterProviderCustomizer(
+                            (meterProviderBuilder, __) -> meterProviderBuilder.registerMetricReader(reader));
+                    builderCustomizer.addPropertiesSupplier(
+                            () -> Map.of(OpenTelemetryService.OTEL_SDK_DISABLED_KEY, "false"));
+                });
+                openTelemetryMetricReader(reader);
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/PulsarTestContext.java:744]: I'm starting to dislike this Lombok :)

[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/PulsarTestContext.java:744]: Ok, I completely missed that this code was inside the builder class

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/BrokerServiceLookupTest.java [hunk 10]
```diff
@@ -178,6 +188,49 @@ public void testMultipleBrokerLookup() throws Exception {
         doReturn(Optional.of(resourceUnit)).when(loadManager2).getLeastLoaded(any(ServiceUnitId.class));
         loadManagerField.set(pulsar.getNamespaceService(), new AtomicReference<>(loadManager1));
 
+        var metricReader = pulsarTestContext.getOpenTelemetryMetricReader();
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/BrokerServiceLookupTest.java:192]: I'm not sure I understand why this won't work.
You hold the latch - making the lookup operation takes as much as you control. You test the metric before and assert it is equal to 0, then run the lookup command using the subscribe(), which will make the lock trigger; at this point, you assert it is equal to 1, and then you release the countDown latch. It seems like two latches are in order.

Scope-wise: I agree that the test is an integration test, so technically, mocking namespace service and spying on a specific method is coupling to implementation. Yet, when I compare mocking a defined service (NamespaceService) method - actually spying on it - versus an internal implementation detail - a semaphore created in broker service, which happens to be used by multiple other classes, I opt for the mocking of namespace service. 

In summary, I'm ok with leaving it as is, but I am not very content with it :)

**Comment by asafm** [comment]
@dragosvictor Thanks. I replied to all comments.

**Review by asafm — COMMENTED** [review]
Also, I liked your idea you explain in the comment you made on 
`pulsar.broker.topic.lookup.operation.pending.[usage,limit]` - having its name align since you said this code is also called outside of AdminAPI.
If this is the case, it makes total sense to align it based on your suggestion.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/common/stats/package-info.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * Provides metric utility classes.
+ */
+package org.apache.pulsar.common.stats;
```
Comment (lhotari) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/common/stats/package-info.java:23]: since this is the `pulsar-broker-common` module, I think that we should use a package under `org.apache.pulsar.broker` to prevent the problem of split packages. 
`pulsar-common` already contains the `org.apache.pulsar.common.stats` package, so that package name is already taken. Split packages aren't allowed in Java Platform Modules System (JPMS, Java 9 modules) either.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 12]
```diff
@@ -175,6 +199,12 @@ public NamespaceService(PulsarService pulsar) {
         this.bundleSplitListeners = new CopyOnWriteArrayList<>();
         this.localBrokerDataCache = pulsar.getLocalMetadataStore().getMetadataCache(LocalBrokerData.class);
         this.redirectManager = new RedirectManager(pulsar);
+
+        this.lookupLatencyHistogram = pulsar.getOpenTelemetry().getMeter()
+                .histogramBuilder("pulsar.broker.lookup.request.duration")
+                .setDescription("Lookup request duration")
+                .setUnit("s")
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:206]: Is there a specific reason to use second as a unit for latency? 
I'd assume that microseconds or nanoseconds would be more useful and avoid unnecessary floating point number operations.

**Review by lhotari — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 13]
```diff
@@ -403,15 +412,43 @@ public BrokerService(PulsarService pulsar, EventLoopGroup eventLoopGroup) throws
...
                 .register();
+        this.pendingLookupOperationsCounter = pulsar.getOpenTelemetry().getMeter()
+                .upDownCounterBuilder("pulsar.broker.topic.lookup.operation.pending.usage")
+                .setDescription("The number of pending lookup operations in the broker. "
+                        + "When it reaches threshold \"maxConcurrentLookupRequest\" defined in broker.conf, "
+                        + "new requests are rejected.")
+                .setUnit("{operation}")
+                .buildWithCallback(measurement -> measurement.record(getPendingLookupRequest()));
+        this.pendingLookupOperationsLimitGauge = pulsar.getOpenTelemetry().getMeter()
+                .gaugeBuilder("pulsar.broker.topic.lookup.operation.pending.limit")
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:425]: I need to figure out why on other examples in OTel they used UpDownCounter. For example: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/runtime/jvm-metrics.md#metric-jvmmemorylimit

[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:425]: Found this: https://opentelemetry.io/docs/specs/otel/metrics/supplementary-guidelines/

[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:425]: I can presume that people may want a cluster-wide sum of topic lookup operations since they see that ZK is under stress and would like to know the cause, which a single broker does not cause. Once you define the pending usage as UpDownCounter, it only makes sense the limit is also UpDownCounter.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/common/stats/package-info.java (hunk 11 above)
Comment (asafm) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/common/stats/package-info.java]: I don't understand the problem

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 12 above)
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:206]: As the the whole world will use those conventions, we better follow them to align.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 14]
```diff
@@ -403,15 +412,43 @@ public BrokerService(PulsarService pulsar, EventLoopGroup eventLoopGroup) throws
...
+                .supplier(this::getPendingTopicLoadRequests)
                 .register();
+        this.pendingTopicLoadOperationsCounter = pulsar.getOpenTelemetry().getMeter()
+                .upDownCounterBuilder("pulsar.broker.topic.load.operation.pending.usage")
+                .setDescription("The number of pending topic load operations in the broker. "
+                        + "When it reaches threshold \"maxConcurrentTopicLoadRequest\" defined in broker.conf, "
+                        + "new requests are rejected.")
+                .setUnit("{operation}")
+                .buildWithCallback(measurement -> measurement.record(getPendingTopicLoadRequests()));
+        this.pendingTopicLoadOperationsLimitGauge = pulsar.getOpenTelemetry().getMeter()
```
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:444]: Keeping this unresolved until deciding

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: I think we can't change it to lookup because multiple topic lookup request can only load one topic. 

Actually, I think it should be 

```
pulsar.broker.topic.concurrent.load.usage
pulsar.broker.topic.concurrent.load.limit
```
We also use `maxConcurrentTopicLoadRequest` in the broker.conf. Keep the name consistent will be better for users. Do not need to translate from concurrent -> pending.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 15]
```diff
@@ -175,6 +187,25 @@ public NamespaceService(PulsarService pulsar) {
...
+        this.lookupFailuresCounter = meter
+                .counterBuilder("pulsar.broker.lookup.failure")
+                .setDescription("The number of lookup failures")
+                .build();
+        this.lookupAnswersCounter = meter
+                .counterBuilder("pulsar.broker.lookup.answer")
+                .setDescription("The number of lookup responses (i.e. not redirected requests)")
+                .build();
+        this.lookupLatencyHistogram = meter
+                .histogramBuilder("pulsar.broker.lookup.latency")
```
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: IMO, we can only have one attribute `response.status`. The url looks too details to the metrics? if we have 50 or 100 brokers.

pulsar.broker.lookup.request.duration(response.status="connect/redirect/failed")

[the remark this answers is not in the document]
Comment (dragosvictor) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: I agree that including the URL provides little value. 

As for the attribute name, we should keep the `pulsar` prefix in place to avoid confusion or clashes with attributes in other namespaces.

I liked @asafm's proposal to split this by a general status (success or failure), as this can further be used by other metrics while also adding details relating to this particular metric where we can (connect/redirect). I think this is an excellent structure to follow going forward.

To reiterate, we'd have a separate histogram record for the following attribute sets (ignoring general resource attributes, such as `pulsar.cluster`):
`pulsar.lookup.response.type=broker,pulsar.response.status=success`
`pulsar.lookup.response.type=redirect,pulsar.response.status=success`
`pulsar.response.status=failure`.

[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: @codelipenghui The attribute name is `pulsar.lookup.response.type`, and it can have the following values: `broker` or `redirect`. We are not placing the actual URL as the attribute value.

Given this, are you ok with the 2 attributes suggested and the name?

Name = `pulsar.broker.lookup.request.duration`
Attributes:
1. `pulsar.lookup.response.type`: values can be `broker`, `redirect`
2. `pulsar.response.status`: values can be `success`, `failure`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: @codelipenghui Couple of questions here:
1. Namespace matters. We try to avoid writing in "English": Topic concurrent load --> `pulsar.broker.topic.concurrent.load"...
   Namespace (prefix) means we might have other metrics under it. I don't see any other metrics for `topic.concurrent`. Hence to me it makes more sense to have `pulsar.broker.topic.load.concurrent` and then `usage` and `limit`, where usage is the current number of concurrent load operations, and limit is the defined limit.
2. Regarding the word `operation` suggested. Let's discuss this.
    Lookup is a request right? It's coming from the user in `/lookup` or in the `lookup` command in the binary protocol. Yet, @dragosvictor mentioned that lookup is actually called from within other requests like get topic of namespaces, get partition metadata, etc. Perhaps the correct thing here as @dragosvictor said is that lookup we measure is actually an operation. 
    Same with lopic load. It is not happening due to a request, but an internal decision, hence it is an operation.
    So perhaps the names should actually be:
   - `pulsar.broker.topic.operations.lookup.concurrent.usage`
   - `pulsar.broker.topic.operations.lookup.concurrent.limit`
   - `pulsar.broker.topic.operations.load.concurrent.usage`
   - `pulsar.broker.topic.operations.load.concurrent.usage`
   
@lhotari @codelipenghui Would love your opinion about this

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: Both `broker, redirect` connote success. Why do we need to have (broker, success), (redirect, success)? For failures, it could be a couple of reason, topic not found or some temporary reason (metadata issue), I'm not sure how the broker can say the reason is belongs to `redirect` or `broker`. So the response type will be absent for the failed lookup response.

https://github.com/apache/pulsar/blob/ccc2ea67c192b4771c9af30c6eaf994fe01958e5/pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java#L974-L982

I think the name defined in Pulsar is little confusing. If the name is LookupResponseCode, it will be easy to comprehend than LookupType :)

IMO, `broker`, `redirect` and `failure` are just like the HTTP response code `200`, `307`, `500`.

Another question about the attribute name. I guess we will have global attributes so we must have `pulsar.lookup` as prefix? Is it possible to have local attribute for `pulsar.lookup.response.type`? It could only work with some metrics related to lookup.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: @asafm 

Oh, I thought the namespace means Pulsar namespace before.

1. So I understand correctly, we will have a couple of metric namespaces:

- pulsar.broker.*
- pulsar.tenant.*
- pulsar.namespace.*
- pulsar.topic.*
- pulsar.subscription.*

Is it correct? And, `pulsar.broker.topic.load.concurrent.*` also looks good to me.

2. But I think it should be covered by metrics for other APIs? For example, deleting a topic will validate the ownership of the topic. The the requested broker is not the owner of the topic, the broker will return 307 to the HTTP client. I think we should have metrics for topic deletion API. For topic lookup, we usually pay more attention to the client lookup requests and the metadata operations from the lookup request (We will skip the metadata operation if the broker cached metadata), not the internal calls to the broker cached metadata (validate topic ownership). And for the metadata operation, we will have metadata driver level metrics.

And, at least we need `lookup.request` for the client side lookup requests. Otherwise, we don't know how many lookup `operation` is from client side lookup requests or internal topic ownership validation.

[the remark this answers is not in the document]
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: perhaps generic parts of the metric name could be removed? just wondering if parts like `concurrent` or `operations` are required at all?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: Ok, after a good discussion with @codelipenghui, with came up with the following instrument names:

```
pulsar.broker.topic.load.duration
pulsar.broker.topic.load.concurrent.usage
pulsar.broker.topic.load.concurrent.limit

pulsar.broker.request.topic.lookup.duration
pulsar.broker.request.topic.lookup.concurrent.usage
pulsar.broker.request.topic.lookup.concurrent.limit
```

Why?
1. The 3rd name is usually a category. For example, messaging, transaction, offloader, etc. After some discussion, we figured "request" could be a decent category for requests like lookup.
   * The code that uses the lookup semaphore needs to be corrected. The only ones who should be using it are the actual lookup HTTP request and the binary command lookup. The rest are wrong, and they should be removed.
2. Internal operations such as topic load can be implicitly assumed, so we've removed them from the instrument naming.

Regarding the attributes, we were thinking of uniting the two attributes into one:
```
	Attributes:
		pulsar.lookup.response = broker, redirect, failure
```

The reasoning was that, like the HTTP status code, pulsar lookup response (code) can signify different responses. We can revisit that as we advance to other requests in the future - we have marked all of this experimental.

@dragosvictor @codelipenghui @lhotari

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java]: See response at https://github.com/apache/pulsar/pull/22058#discussion_r1511099159

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (dragosvictor) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: Name changes sound good. Note that there is no metric `pulsar.broker.topic.load.duration
` as of right now, I'm not sure why it got included in the list.

[the remark this answers is not in the document]
Comment (asafm) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java]: We were brainstorming many metrics :)

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java [hunk 16]
```diff
@@ -201,7 +201,7 @@ public NamespaceService(PulsarService pulsar) {
         this.redirectManager = new RedirectManager(pulsar);
 
         this.lookupLatencyHistogram = pulsar.getOpenTelemetry().getMeter()
-                .histogramBuilder("pulsar.broker.lookup.request.duration")
+                .histogramBuilder(LOOKUP_REQUEST_DURATION_METRIC_NAME)
```
Comment (asafm) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:204]: The description should match how OTel writes it: "The duration of topic lookup requests (either binary or HTTP)". @codelipenghui Am I right?

Comment (dragosvictor) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:204]: Done!

**Review by asafm — APPROVED** [review]
💯

**Comment by asafm** [comment]
@lhotari You're up next on the review wrap up :)

**Review by dao-jun — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java (hunk 12 above)
Comment (lhotari) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java:206]: Thank you @dragosvictor for the pointer. I also found https://github.com/open-telemetry/opentelemetry-specification/issues/2977 which was helpful as background information.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/common/stats/package-info.java (hunk 11 above)
Comment (lhotari) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/common/stats/package-info.java]: Couldn't we address this package split issue already in this PR before proceeding?

Comment (lhotari) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/common/stats/package-info.java]: Oh I see that no new split package anymore in this PR, problem resolved.

**Review by lhotari — APPROVED** [review]
LGTM. Good work @dragosvictor
